xclim/core/base.py

```diff
@@ -0,0 +1,407 @@
+
+    Climate indicator object that, when called, computes an indicator and assigns its output a number of
+    CF-compliant attributes. Some of these attributes can be *templated*, allowing metadata to reflect
+    the value of call arguments.
+
+    Instantiating a new indicator returns an instance but also creates and registers a custom subclass.
+
+    Parameters
+    ----------
+    identifier: str
+      Unique ID for class registry, should be a valid slug.
+    var_name: str
+      Output variable name. May use tags {<tag>}.
+    standard_name: str
+      Variable name (CF).
+    long_name: str
+      Descriptive variable name.
+    units: str
+      Representative units of the physical quantity (CF).
+    compute: func
+      The function computing the indicator.
+    cell_methods: str
+      List of blank-separated words of the form "name: method" (CF).
+    description: str
+      Sentence meant to clarify the qualifiers of the fundamental quantities, such as which
+      surface a quantity is defined on or what the flux sign conventions are.
+    context: str
+      The `pint` unit context, for example use 'hydro' to allow conversion from kg m-2 s-1 to mm/day.
+    title: str, None
+      A succinct description of what is in the computed output. Parsed from `compute` docstring if None.
+    abstract: str
+      A long description of what is in the computed output. Parsed from `compute` docstring if None.
+    keywords: str
+      Comma separated list of keywords. Parsed from `compute` docstring if None.
+    references: str
+      Published or web-based references that describe the data or methods used to produce it. Parsed from
+      `compute` docstring if None.
+    comment: str
+      Miscellaneous information about the data or methods used to produce it.
+    notes: str
+      Notes regarding computing function, for example the mathematical formulation. Parsed from `compute`
+      docstring if None.
+
+    Notes
+    -----
+    All subclasses created are available in the `registry` attribute and can be used to defined custom subclasses.
+
+    """
+    # Number of DataArray variables
+    _nvar = 1
+
+    # Allowed metadata attributes on the output
+    _cf_names = [
+        "standard_name",
+        "long_name",
+        "units",
+        "cell_methods",
+        "description",
+        "comment",
+        "references",
+    ]
+
+    # metadata fields that are formatted as free text.
+    _text_fields = ["long_name", "description", "comment"]
+
+    _funcs = ["compute", "missing_func", "cfprobe", "validate"]
+
+    # Default attribute values
+    identifier = None
+    var_name = None
+    standard_name = ""
+    long_name = ""
+    units = ""
+    missing_func = staticmethod(checks.missing_any)
+    cell_methods = ""
+    description = ""
+    context = "none"
+    title = ""
+    abstract = ""
+    keywords = ""
+    references = ""
+    comment = ""
+    notes = ""
+
+    # Subclass registry
+    registry = {}
```

Comment:
Is there a reason for having the registry as an attribute of the Indicator class, instead of the module variable?

Reply:
To differentiate descendants. So Indicator2D has its own registry. Open to changing this, or registering subclasses at the module level and class level.

Reply:
Is the Indicator vs Indicator2D the only division? Honestly, I don't really understand the need of having a subclass of Indicator just to change `_nvar`, why not simply change it in the subclass itself? Or even leave it to the object instantiation. That's what I did in  `_conversion.py`, it didn't seem useful to create many subclasses only to change `_nvar`.

Also, if we restrict input arguments to DataArrays and parameters to numbers or strings, then we could easily infer `nvar` from the annotations of `compute`?

Reply:
This may have been done to facilitate wrapping in the WPS server. We can probably work something out without this distinction.